scripts/test_analysis/main.go

```diff
@@ -0,0 +1,226 @@
+				if _, ok := loggedFlaky[flakyName]; !ok {
+					loggedFlaky[flakyName] = struct{}{}
+					log.Printf("Test %s.%s is flaky.", ft.Package, ft.Test)
+				}
+			}
+		}
+		if !isFlaky {
+			hasOneNonFlakyFailure = true
+		}
+	}
+
+	// A test consistently failed, so we should exit with a non-zero exit code.
+	if hasOneNonFlakyFailure {
+		return errors.New("one or more tests consistently failed")
+	}
+	return nil
+}
+
+func (t *tester) goTestAll(extraFlags []string) error {
+	flags := []string{"./..."}
+	flags = append(flags, extraFlags...)
+	return t.goTest(flags)
+}
+
+func (t *tester) goTestPkgTest(pkg, testname string, extraFlags []string) error {
+	flags := []string{
+		pkg, "-run", "^" + testname + "$", "-count", "1",
+	}
+	flags = append(flags, extraFlags...)
+	return t.goTest(flags)
+}
+
+func (t *tester) goTest(extraFlags []string) error {
+	flags := []string{
+		"test", "-json",
+	}
+	flags = append(flags, extraFlags...)
+	cmd := exec.Command("go", flags...)
+	cmd.Dir = t.Dir
+	cmd.Stderr = os.Stderr
+
+	gotest2sql := exec.Command("gotest2sql", "-v", "-output", dbPath)
+	gotest2sql.Dir = t.Dir
+	gotest2sql.Stdin, _ = cmd.StdoutPipe()
+	gotest2sql.Stdout = os.Stdout
+	gotest2sql.Stderr = os.Stderr
+	err := gotest2sql.Start()
+	if err != nil {
+		return err
+	}
+
+	err = cmd.Run()
+	return errors.Join(err, gotest2sql.Wait())
+}
+
+type failedTest struct {
+	Package string
+	Test    string
+}
+
+func (t *tester) findFailedTests(ctx context.Context) ([]failedTest, error) {
+	db, err := sql.Open("sqlite", t.Dir+dbPath)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+
+	rows, err := db.QueryContext(ctx, "SELECT DISTINCT Package, Test FROM test_results where Action='fail' and Test != ''")
+	if err != nil {
+		return nil, err
+	}
+	var out []failedTest
+	for rows.Next() {
+		var pkg, test string
+		if err := rows.Scan(&pkg, &test); err != nil {
+			return nil, err
+		}
+		out = append(out, failedTest{pkg, test})
+	}
+	return out, nil
+}
+
+func filterOutFlags(flags []string, exclude *regexp.Regexp) []string {
+	out := make([]string, 0, len(flags))
+	for _, f := range flags {
+		if !exclude.MatchString(f) {
+			out = append(out, f)
+		}
+	}
+	return out
+}
+
+// summarize returns a markdown string of the test results.
+func (t *tester) summarize() (string, error) {
+	ctx := context.Background()
+	var out strings.Builder
+
+	testFailures, err := t.findFailedTests(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	plural := "s"
+	if len(testFailures) == 1 {
+		plural = ""
+	}
+	out.WriteString(fmt.Sprintf("## %d Test Failure%s\n\n", len(testFailures), plural))
+
+	db, err := sql.Open("sqlite", t.Dir+dbPath)
+	if err != nil {
+		return "", err
+	}
+	defer db.Close()
+
+	rows, err := db.QueryContext(ctx, `SELECT
+    tr_output.Package,
+    tr_output.Test,
+    GROUP_CONCAT(tr_output.Output,  "") AS Outputs
+FROM
+    test_results tr_fail
+JOIN
+    test_results tr_output
+ON
+    tr_fail.Test = tr_output.Test
```

Comment:
Why do we need a self join here? Can't we just do a GROUP BY?

Reply:
same reason as https://github.com/libp2p/go-libp2p/pull/2923#discussion_r1723675521. Each row is just a different state/output line from `go test`. A single row doesn't represent the whole state of a single test.